enhancements/network/routed-ingress-primary-udn-static-ips.md

```diff
@@ -0,0 +1,565 @@
+a VM into OpenShift with it's MAC, IPs, and gateway on top of primary UDN is the
+missing piece we should pursue.
+
+## Motivation
+
+Some users are running VMs in virtualization platforms having a managed IP
+configuration.
+
+For this kind of users, we need to have a way to enable their existing VMs to
+run properly after being migrated into OpenShift, without any guest
+configuration changes. For that, we need to import the VMs from their existing
+platforms, preserving their existing MACs, IPs, and gateway configuration.
+
+### User Stories
+
+- As the owner of a VM running in a traditional virtualization platform, I want
+to import said VM - whose IPs were statically configured - into Kubernetes,
+attaching it to an overlay network. I want to ingress/egress using the same IP
+address (the one assgined to the VM).
+- As the owner of a VM running in a traditional virtualization platform, I want
+to import said VM into Kubernetes, attaching it to an overlay network. I want to
+have managed IPs for my VMs, since that feature was already available in my
+previous platform.
+- As the owner of a VM running in a traditional virtualization platform, I want
+to import said VM into Kubernetes, attaching it to an overlay network. I want to
+be able to consume Kubernetes features like network policies, and services, to
+benefit from the Kubernetes experience.
+
+### Goals
+
+- Preserve the original VM's MAC address
+- Preserve the original VM's IP address
+- Specify the gateway IP address of the imported VM so it can keep the same
+default route
+- Allow excludeSubnets to be used with L2 UDNs to ensure OVNK does not use an IP
+address from the range that VMs have already been assigned outside of the
+cluster (or for secondary IP addresses assigned to the VM's interfaces)
+
+### Non-Goals
+
+Handle importing VMs without a managed IP experience - i.e. IPs were defined
+statically in the guest.
+
+## Proposal
+
+This proposal to import VMs from other virtualization platforms while
+preserving their networking configuration will impact three different OpenShift
+components:
+- MTV
+- KubeVirt's ipam-extensions
+- OVN-Kubernetes
+
+We will elaborate the overall flow first, before digging into details in each
+of the components. There will be two sub-sections: one about preserving the
+[IP addresses](#preserving-the-original-vm-ip-address),
+another about preserving the
+[gateway configuration](#preserving-the-vm-gateway).
+MTV already introspects the VM on the source cluster, and templates the VM with
+its original MAC addresses.
+
+### Preserving the original VM IP address
+
+MTV currently introspects the VM to learn the MAC addresses of the interfaces,
+and templates the VM which will be created in OpenShift Virtualization with the
+required MAC address. We would need MTV to somehow also figure out what IP
+addresses are on the aforementioned interfaces.
+
+A new CRD - named `IPPool`, or `DHCPLeaseConfig` (or the like) - will be
```

Comment:
hmm why not make IPAMClaim's v2alpha2 API `AddressClaims` so enhance IPAMClaim ? :D

Reply:
1. that's not a centralized place with the MAC => IPs association. There's value in that. Plus, it's closer to what existing platforms provide.
2. I feel we'd do that because "there's a box we're opening already. Let's put this disjoint information into the same box, and retrieve both". I.e. we're bending what we have so we reach the goal we think we have faster. 

Having said that, re-purposing IPAMClaims to something more generic **is** an option. I just feel we're not doing it for the right reasons. I'm afraid we might be doing it because it fits the time-frame rather than the use case.

FWIW, the issue of "importing 5000" VMs is still there if we take this route. How will the 5000 `AddressClaims` be provisioned ? By whom ? With what data ?